### ISSUE TITLE
[Android] Context manipulation is working incorrectly in Gmail app

### ISSUE THREAD

**Issue by MakaraSok** [keymanapp/keyman#2250]
Originally from [community site forum](https://community.software.sil.org/t/rotas-eat-characters-in-android-gmail-app/2713)

**Describe the bug**
The character rotation feature does not give the expected outputs. It instead deletes character (s) before the character which has to be rotated leaving incomplete string of characters, i.e. gibberish.

**To Reproduce**
1. Go to 'Gmail app'
2. Click to compose a new email
3. Start to type the character in the following orders using Khmer Angkor keyboard.
- (Type: `sea`)  ស េ ា > ោ not the expected សោ
- (Type: `seu`) ស  េ ុ > ុ not សុ  
- (Type: `suI`)  សុ ី > ស៊ី
- (Type `kNjd`) ក ណ ្ត > ណ្ដ not កណ្ដ
- (Type `xEjm`) ខ ែ ្ម > ្ម not ខ្មែ

**Expected behavior**
The rotation should not delete any character which does not belong in the rules. 

**Screenshots**
https://youtu.be/MpU7VAmhpB8

---

  
**Android:**
 - Device: [Galaxy S9+]
 - OS: [Android 9.0]
 - Keyman version [12.0.4205]
 - Target application: [Gmail for Android version 2019.09.15.270135155.release]
 
---
 
**Keyboard**
 - Keyboard name: [khmer_angkor]
 - Keyboard version: [1.0.1]
 - Language name: [Khmer]

**Additional context**
* I've tested this behavior in Google Keep, and it works fine there.
* Typing in Gmail subject field works fine

**Comment by darcywong00** [comment]
Minimal keyboard for repro
```
store(&NAME) 'moo'
store(&VISUALKEYBOARD) 'moo.kvks'
store(&TARGETS) 'windows mobile'
store(&LAYOUTFILE) 'moo.keyman-touch-layout'
begin Unicode > use(main)

group(main) using keys

'p' + 'p' > 'ṗ'

'ṗ' + 'p' > 'p̂'

'p̂' + 'p' > 'p̂p̂'

'p̂p̂' + 'p' > 'p'
```

Link to [moo.kmp](http://darcywong00.github.io/examples/moo/moo.kmp)

**Comment by darcywong00** [comment]
### Using gmail webview (Chrome browser) on the Android device

In all these fields, the rotas eventually delete characters:
* To:
* Cc/Bcc:
* Subject:
* Body:

**Comment by darcywong00** [comment]
Additional info from more investigation

### Test Scenario

* I can replicate the issue on an input field in Chrome
```
<form>
  <input type='text'>
</form>
``` 

so it's not isolated to Gmail.
Note: 
`ṗ` is two code points 
`p`   0070 Latin small latter p
`̇`     0307 combining dot above

`p̂` is two code points
`p` 0070 Latin small letter p
`̂`  0302 combining circumflex accent

Context | Type | Expected | Actual
|--------|--------|-----------|--------|
|            |    o      |      o       |     o    |
|   o       |     p     |      op     |   op    |
|  op      |     p     |     oṗ      |   oṗ    |
|   oṗ     |      p    |     op̂       |    p̂     |



* Issue existed in Keyman Android 10 and 11, so it's not lm-layer 
* Using the keyboard on a KMW test page does not have the issue

**Comment by mcdurdin** [comment]
This is definitely a [bug in Chromium](https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/editing/ime/input_method_controller.cc?type=cs&q=InputMethodController::deleteSurroundingText&sq=package:chromium&g=0&l=1237&targetos=android). This has been documented as a TODO but not [listed as a bug](https://bugs.chromium.org/p/chromium/issues/list?q=deleteSurroundingText&can=1) by Chromium as of issue date.

However, we need a workaround because it will take a long time for this fix, if and when it lands, to filter down to all the devices we support.

Proposal:

1. Retrieve the text before cursor, `CharSequence charsBackup = ic.getTextBeforeCursor(dn*2 + 16, 0);` into a buffer. This ensures we collect enough characters for surrogate pairs + a long grapheme cluster at the start of the text.
2. Count the number of surrogate pairs in this buffer, backwards from the end until we reach `dn` codepoints (remember `dn` is in [code points, not code units](https://stackoverflow.com/a/27331885/1836776)). 
3. Update `charsBackup` to delete `dn+numPairs` from the right of the buffer.
3. Now tell the app to do the same (`ic.deleteSurroundingText(dn+numPairs, 0);`).
4. Now we run into the Chromium bug: it may have deleted more than we want it to, all the way up to a grapheme cluster boundary. The good news is we should just be able to put it back again because we've kept a copy in `charsBackup`.
5. So, `getTextBeforeCursor()` again. Slide this new buffer left until it matches the text in `charsBackup`. The dangling characters in `charsBackup` should be prepended to the text to be inserted in `ic.commitText` further down.

Does that make sense? There's some detail I've elided there but hoping that's enough to get teeth into it!

Finally, we should raise this as a bug with Chromium as well.
